hostverify library: Cannot use oe_verify_attestation_certificate_with_evidence

This is with current `master`:
```
../openenclave/lib/openenclave/host/liboehostverify.a(verifier.c.o): In function `_verify_report':
.../common/sgx/verifier.c:742: undefined reference to `oe_verify_report_internal'
```
Looks like some sources are missing in the archive.

--- a/host/CMakeLists.txt
+++ b/host/CMakeLists.txt
@@ -210,6 +210,7 @@ if (OE_SGX)
     ../common/sgx/tlsverifier.c
     ../common/sgx/verifier.c
     sgx/hostverify_report.c
+    sgx/report_common.c
     sgx/sgxquoteprovider.c)
 
   if (WITH_EEID)
@@ -233,6 +234,7 @@ if (OE_SGX)
     sgx/ocalls/thread.c
     sgx/quote.c
     sgx/registers.c
+    sgx/report_common.c
     sgx/report.c
     sgx/sgx_enclave_common_wrapper.c
     sgx/sgxload.c
@@ -469,6 +471,8 @@ target_compile_definitions(
   oehostverify
   PRIVATE $<TARGET_PROPERTY:oehost,COMPILE_DEFINITIONS>
   INTERFACE $<TARGET_PROPERTY:oehost,INTERFACE_COMPILE_DEFINITIONS>)
+
+target_compile_definitions(oehostverify PUBLIC OE_BUILD_HOST_VERIFY)
 
 # TODO: Remove this hard coded output directory.
 set_property(TARGET oehost PROPERTY ARCHIVE_OUTPUT_DIRECTORY

--- a/host/sgx/report.c
+++ b/host/sgx/report.c
@@ -2,168 +2,13 @@
 // Licensed under the MIT License.
 
 #include <openenclave/attestation/sgx/evidence.h>
-#include <openenclave/host.h>
-#include <openenclave/host_verify.h>
-#include <openenclave/internal/calls.h>
 #include <openenclave/internal/plugin.h>
 #include <openenclave/internal/raise.h>
 #include <openenclave/internal/report.h>
-#include <openenclave/internal/safecrt.h>
-#include <openenclave/internal/safemath.h>
-#include <openenclave/internal/sgx/plugin.h>
-#include <openenclave/internal/trace.h>
-#include <openenclave/internal/utils.h>
-#include "../common/sgx/quote.h"
-#include "core_u.h"
+
 #include "platform_u.h"
-#include "quote.h"
-
-#include "sgxquoteprovider.h"
-
-/**
- * Declare the prototypes of the following functions to avoid the
- * missing-prototypes warning.
- */
-oe_result_t _oe_get_report_v2_ecall(
-    oe_enclave_t* enclave,
-    oe_result_t* _retval,
-    uint32_t flags,
-    const void* opt_params,
-    size_t opt_params_size,
-    uint8_t** report_buffer,
-    size_t* report_buffer_size);
-oe_result_t _oe_verify_local_report_ecall(
-    oe_enclave_t* enclave,
-    oe_result_t* _retval,
-    const uint8_t* report,
-    size_t report_size,
-    oe_report_t* parsed_report);
-oe_result_t _oe_verify_report_ecall(
-    oe_enclave_t* enclave,
-    oe_result_t* _retval,
-    const void* report,
-    size_t report_size);
-
-/**
- * Make the following ECALLs weak to support the system EDL opt-in.
- * When the user does not opt into (import) the EDL, the linker will pick
- * the following default implementations. If the user opts into the EDL,
- * the implementations (which are strong) in the oeedger8r-generated code will
- * be used. This behavior is guaranteed by the linker; i.e., the linker will
- * pick the symbols defined in the object before those in the library.
- */
-oe_result_t _oe_get_report_v2_ecall(
-    oe_enclave_t* enclave,
-    oe_result_t* _retval,
-    uint32_t flags,
-    const void* opt_params,
-    size_t opt_params_size,
-    uint8_t** report_buffer,
-    size_t* report_buffer_size)
-{
-    OE_UNUSED(enclave);
-    OE_UNUSED(flags);
-    OE_UNUSED(opt_params);
-    OE_UNUSED(opt_params_size);
-    OE_UNUSED(report_buffer);
-    OE_UNUSED(report_buffer_size);
-
-    if (_retval)
-        *_retval = OE_UNSUPPORTED;
-
-    return OE_UNSUPPORTED;
-}
-OE_WEAK_ALIAS(_oe_get_report_v2_ecall, oe_get_report_v2_ecall);
-
-oe_result_t _oe_verify_local_report_ecall(
-    oe_enclave_t* enclave,
-    oe_result_t* _retval,
-    const uint8_t* report,
-    size_t report_size,
-    oe_report_t* parsed_report)
-{
-    OE_UNUSED(enclave);
-    OE_UNUSED(report);
-    OE_UNUSED(report_size);
-    OE_UNUSED(parsed_report);
-
-    if (_retval)
-        *_retval = OE_UNSUPPORTED;
-
-    return OE_UNSUPPORTED;
-}
-OE_WEAK_ALIAS(_oe_verify_local_report_ecall, oe_verify_local_report_ecall);
-
-oe_result_t _oe_verify_report_ecall(
-    oe_enclave_t* enclave,
-    oe_result_t* _retval,
-    const void* report,
-    size_t report_size)
-{
-    OE_UNUSED(enclave);
-    OE_UNUSED(report);
-    OE_UNUSED(report_size);
-
-    if (_retval)
-        *_retval = OE_UNSUPPORTED;
-
-    return OE_UNSUPPORTED;
-}
-OE_WEAK_ALIAS(_oe_verify_report_ecall, oe_verify_report_ecall);
-
-OE_STATIC_ASSERT(OE_REPORT_DATA_SIZE == sizeof(sgx_report_data_t));
 
 static const oe_uuid_t _ecdsa_uuid = {OE_FORMAT_UUID_SGX_ECDSA_P256};
-
-oe_result_t oe_get_report_v2(
-    oe_enclave_t* enclave,
-    uint32_t flags,
-    const void* opt_params,
-    size_t opt_params_size,
-    uint8_t** report_buffer,
-    size_t* report_buffer_size)
-{
-    oe_result_t result = OE_UNEXPECTED;
-    uint8_t* report = NULL;
-    size_t report_size = 0;
-
-    if (!enclave || !report_buffer || !report_buffer_size)
-        OE_RAISE(OE_INVALID_PARAMETER);
-
-    OE_CHECK(oe_get_report_v2_ecall(
-        enclave,
-        &result,
-        flags,
-        opt_params,
-        opt_params_size,
-        &report,
-        &report_size));
-    OE_CHECK(result);
-
-    *report_buffer = report;
-    *report_buffer_size = report_size;
-    report = NULL;
-
-    result = OE_OK;
-
-done:
-    if (report)
-        oe_free_report(report);
-
-    if (result == OE_UNSUPPORTED)
-        OE_TRACE_WARNING(
-            "SGX remote attestation is not enabled. To enable, please add\n\n"
-            "from \"openenclave/edl/sgx/attestation.edl\" import *;\n\n"
-            "in the edl file.\n");
-
-    return result;
-}
-
-void oe_free_report(uint8_t* report_buffer)
-{
-    free(report_buffer);
-}
-
 // Host version, supports ECDSA remote attestation natively.
 // for SGX local attestation, it makes ecall to the enclave.
 oe_result_t oe_verify_report(
@@ -217,63 +62,46 @@ done:
     return result;
 }
 
-oe_result_t oe_verify_report_internal(
+oe_result_t oe_get_report_v2(
     oe_enclave_t* enclave,
-    const uint8_t* report,
-    size_t report_size,
-    oe_report_t* parsed_report)
+    uint32_t flags,
+    const void* opt_params,
+    size_t opt_params_size,
+    uint8_t** report_buffer,
+    size_t* report_buffer_size)
 {
     oe_result_t result = OE_UNEXPECTED;
-    oe_report_t oe_report = {0};
-    oe_report_header_t* header = (oe_report_header_t*)report;
+    uint8_t* report = NULL;
+    size_t report_size = 0;
 
-    if (report == NULL)
+    if (!enclave || !report_buffer || !report_buffer_size)
         OE_RAISE(OE_INVALID_PARAMETER);
 
-    if (report_size == 0 || report_size > OE_MAX_REPORT_SIZE)
-        OE_RAISE(OE_INVALID_PARAMETER);
+    OE_CHECK(oe_get_report_v2_ecall(
+        enclave,
+        &result,
+        flags,
+        opt_params,
+        opt_params_size,
+        &report,
+        &report_size));
+    OE_CHECK(result);
 
-    // Ensure that the report is parseable before using the header.
-    OE_CHECK(oe_parse_report(report, report_size, &oe_report));
-
-    if (header->report_type == OE_REPORT_TYPE_SGX_REMOTE)
-    {
-        // Intialize the quote provider if we want to verify a remote quote.
-        // Note that we don't need the sgx libraries to verify the quote. All we
-        // need is the quote provider.
-        OE_CHECK(oe_initialize_quote_provider());
-
-        // Quote attestation can be done entirely on the host side.
-        OE_CHECK(oe_verify_sgx_quote(
-            header->report, header->report_size, NULL, 0, NULL));
-    }
-    else if (header->report_type == OE_REPORT_TYPE_SGX_LOCAL)
-    {
-        oe_result_t retval;
-
-        if (enclave == NULL)
-            OE_RAISE(OE_INVALID_PARAMETER);
-
-        OE_CHECK(oe_verify_report_ecall(enclave, &retval, report, report_size));
-        OE_CHECK(retval);
-    }
-    else
-    {
-        OE_RAISE(OE_INVALID_PARAMETER);
-    }
-
-    // Optionally return parsed report.
-    if (parsed_report != NULL)
-        OE_CHECK(oe_parse_report(report, report_size, parsed_report));
+    *report_buffer = report;
+    *report_buffer_size = report_size;
+    report = NULL;
 
     result = OE_OK;
+
 done:
+    if (report)
+        oe_free_report(report);
+
     if (result == OE_UNSUPPORTED)
-        OE_TRACE_WARNING("oe_verify_report_ecall is not supported. To "
-                         "enable, please add\n\n"
-                         "from \"openenclave/edl/attestation.edl\" import "
-                         "oe_verify_report_ecall;\n\n"
-                         "in the edl file.\n");
+        OE_TRACE_WARNING(
+            "SGX remote attestation is not enabled. To enable, please add\n\n"
+            "from \"openenclave/edl/sgx/attestation.edl\" import *;\n\n"
+            "in the edl file.\n");
 
     return result;
 }

--- a/host/sgx/report_common.c
+++ b/host/sgx/report_common.c
@@ -1,0 +1,183 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/host.h>
+#include <openenclave/host_verify.h>
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/plugin.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/report.h>
+#include <openenclave/internal/safecrt.h>
+#include <openenclave/internal/safemath.h>
+#include <openenclave/internal/sgx/plugin.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+#include "../common/sgx/quote.h"
+
+#if !defined(OE_BUILD_HOST_VERIFY)
+#include "core_u.h"
+#include "platform_u.h"
+#endif
+
+#include "quote.h"
+#include "sgxquoteprovider.h"
+
+/**
+ * Declare the prototypes of the following functions to avoid the
+ * missing-prototypes warning.
+ */
+oe_result_t _oe_get_report_v2_ecall(
+    oe_enclave_t* enclave,
+    oe_result_t* _retval,
+    uint32_t flags,
+    const void* opt_params,
+    size_t opt_params_size,
+    uint8_t** report_buffer,
+    size_t* report_buffer_size);
+oe_result_t _oe_verify_local_report_ecall(
+    oe_enclave_t* enclave,
+    oe_result_t* _retval,
+    const uint8_t* report,
+    size_t report_size,
+    oe_report_t* parsed_report);
+oe_result_t _oe_verify_report_ecall(
+    oe_enclave_t* enclave,
+    oe_result_t* _retval,
+    const void* report,
+    size_t report_size);
+
+/**
+ * Make the following ECALLs weak to support the system EDL opt-in.
+ * When the user does not opt into (import) the EDL, the linker will pick
+ * the following default implementations. If the user opts into the EDL,
+ * the implementations (which are strong) in the oeedger8r-generated code will
+ * be used. This behavior is guaranteed by the linker; i.e., the linker will
+ * pick the symbols defined in the object before those in the library.
+ */
+oe_result_t _oe_get_report_v2_ecall(
+    oe_enclave_t* enclave,
+    oe_result_t* _retval,
+    uint32_t flags,
+    const void* opt_params,
+    size_t opt_params_size,
+    uint8_t** report_buffer,
+    size_t* report_buffer_size)
+{
+    OE_UNUSED(enclave);
+    OE_UNUSED(flags);
+    OE_UNUSED(opt_params);
+    OE_UNUSED(opt_params_size);
+    OE_UNUSED(report_buffer);
+    OE_UNUSED(report_buffer_size);
+
+    if (_retval)
+        *_retval = OE_UNSUPPORTED;
+
+    return OE_UNSUPPORTED;
+}
+OE_WEAK_ALIAS(_oe_get_report_v2_ecall, oe_get_report_v2_ecall);
+
+oe_result_t _oe_verify_local_report_ecall(
+    oe_enclave_t* enclave,
+    oe_result_t* _retval,
+    const uint8_t* report,
+    size_t report_size,
+    oe_report_t* parsed_report)
+{
+    OE_UNUSED(enclave);
+    OE_UNUSED(report);
+    OE_UNUSED(report_size);
+    OE_UNUSED(parsed_report);
+
+    if (_retval)
+        *_retval = OE_UNSUPPORTED;
+
+    return OE_UNSUPPORTED;
+}
+OE_WEAK_ALIAS(_oe_verify_local_report_ecall, oe_verify_local_report_ecall);
+
+oe_result_t _oe_verify_report_ecall(
+    oe_enclave_t* enclave,
+    oe_result_t* _retval,
+    const void* report,
+    size_t report_size)
+{
+    OE_UNUSED(enclave);
+    OE_UNUSED(report);
+    OE_UNUSED(report_size);
+
+    if (_retval)
+        *_retval = OE_UNSUPPORTED;
+
+    return OE_UNSUPPORTED;
+}
+OE_WEAK_ALIAS(_oe_verify_report_ecall, oe_verify_report_ecall);
+
+OE_STATIC_ASSERT(OE_REPORT_DATA_SIZE == sizeof(sgx_report_data_t));
+
+void oe_free_report(uint8_t* report_buffer)
+{
+    free(report_buffer);
+}
+
+oe_result_t oe_verify_report_internal(
+    oe_enclave_t* enclave,
+    const uint8_t* report,
+    size_t report_size,
+    oe_report_t* parsed_report)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_report_t oe_report = {0};
+    oe_report_header_t* header = (oe_report_header_t*)report;
+
+    if (report == NULL)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (report_size == 0 || report_size > OE_MAX_REPORT_SIZE)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    // Ensure that the report is parseable before using the header.
+    OE_CHECK(oe_parse_report(report, report_size, &oe_report));
+
+    if (header->report_type == OE_REPORT_TYPE_SGX_REMOTE)
+    {
+        // Intialize the quote provider if we want to verify a remote quote.
+        // Note that we don't need the sgx libraries to verify the quote. All we
+        // need is the quote provider.
+        OE_CHECK(oe_initialize_quote_provider());
+
+        // Quote attestation can be done entirely on the host side.
+        OE_CHECK(oe_verify_sgx_quote(
+            header->report, header->report_size, NULL, 0, NULL));
+    }
+#ifndef OE_BUILD_HOST_VERIFY
+    else if (header->report_type == OE_REPORT_TYPE_SGX_LOCAL)
+    {
+        if (enclave == NULL)
+            OE_RAISE(OE_INVALID_PARAMETER);
+        oe_result_t retval = OE_OK;
+        OE_CHECK(oe_verify_report_ecall(enclave, &retval, report, report_size));
+        OE_CHECK(retval);
+    }
+#endif
+    else
+    {
+        OE_UNUSED(enclave);
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    // Optionally return parsed report.
+    if (parsed_report != NULL)
+        OE_CHECK(oe_parse_report(report, report_size, parsed_report));
+
+    result = OE_OK;
+done:
+    if (result == OE_UNSUPPORTED)
+        OE_TRACE_WARNING("oe_verify_report_ecall is not supported. To "
+                         "enable, please add\n\n"
+                         "from \"openenclave/edl/attestation.edl\" import "
+                         "oe_verify_report_ecall;\n\n"
+                         "in the edl file.\n");
+
+    return result;
+}

--- a/samples/host_verify/host.c
+++ b/samples/host_verify/host.c
@@ -2,6 +2,7 @@
 // Licensed under the MIT License.
 
 #include <errno.h>
+#include <openenclave/attestation/verifier.h>
 #include <openenclave/host_verify.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -130,27 +131,50 @@ oe_result_t verify_report(
     return result;
 }
 
-oe_result_t enclave_identity_verifier(oe_identity_t* identity, void* arg)
+oe_result_t sgx_enclave_claims_verifier(
+    oe_claim_t* claims,
+    size_t claims_length,
+    void* arg)
 {
+    oe_result_t result = OE_VERIFY_FAILED;
+
     (void)arg;
+    printf("sgx_enclave_claims_verifier is called with claims:\n");
 
-    printf(
-        "Enclave certificate contains the following identity information:\n");
-    printf("identity.security_version = %d\n", identity->security_version);
+    for (size_t i = 0; i < claims_length; i++)
+    {
+        oe_claim_t* claim = &claims[i];
+        if (strcmp(claim->name, OE_CLAIM_SECURITY_VERSION) == 0)
+        {
+            uint32_t security_version = *(uint32_t*)(claim->value);
+            // Check the enclave's security version
+            if (security_version < 1)
+            {
+                printf(
+                    "identity->security_version checking failed (%d)\n",
+                    security_version);
+                goto done;
+            }
+        }
+        // Dump an enclave's unique ID, signer ID and Product ID. They are
+        // MRENCLAVE, MRSIGNER and ISVPRODID for SGX enclaves. In a real
+        // scenario, custom id checking should be done here
+        else if (
+            strcmp(claim->name, OE_CLAIM_SIGNER_ID) == 0 ||
+            strcmp(claim->name, OE_CLAIM_UNIQUE_ID) == 0 ||
+            strcmp(claim->name, OE_CLAIM_PRODUCT_ID) == 0)
+        {
+            printf("Enclave %s:\n", claim->name);
+            for (size_t j = 0; j < claim->value_size; j++)
+            {
+                printf("0x%0x ", claim->value[j]);
+            }
+        }
+    }
 
-    printf("identity->unique_id:\n0x ");
-    for (int i = 0; i < 32; i++)
-        printf("%0x ", (uint8_t)identity->unique_id[i]);
-
-    printf("\nidentity->signer_id:\n0x ");
-    for (int i = 0; i < 32; i++)
-        printf("%0x ", (uint8_t)identity->signer_id[i]);
-
-    printf("\nidentity->product_id:\n0x ");
-    for (int i = 0; i < 16; i++)
-        printf("%0x ", (uint8_t)identity->product_id[i]);
-
-    return OE_OK;
+    result = OE_OK;
+done:
+    return result;
 }
 
 oe_result_t verify_cert(const char* filename)
@@ -161,8 +185,9 @@ oe_result_t verify_cert(const char* filename)
 
     if (read_binary_file(filename, &cert_data, &cert_file_size))
     {
-        result = oe_verify_attestation_certificate(
-            cert_data, cert_file_size, enclave_identity_verifier, NULL);
+        oe_verifier_initialize();
+        result = oe_verify_attestation_certificate_with_evidence(
+            cert_data, cert_file_size, sgx_enclave_claims_verifier, NULL);
     }
 
     if (cert_data != NULL)

